package/ota_daemon/src/ota_daemon.c

```diff
@@ -0,0 +1,529 @@
+  char current[64];
+  device_fw_version_get(current, sizeof(current));
+
+  ota_version_t offered_parsed = {0};
+  if (ota_parse_version_str(offered, &offered_parsed)) {
+    return -1;
+  }
+
+  ota_version_t current_parsed = {0};
+  if (ota_parse_version_str(current, &current_parsed)) {
+    return -1;
+  }
+
+  int ret = ota_cmp_version(&offered_parsed, &current_parsed);
+
+  if (ret > 0) {
+    piksi_log(LOG_INFO, "OTA new version available: %s", offered);
+  }
+
+  return ret;
+}
+
+static int ota_parse_response(ota_resp_t *parsed_resp) {
+  /* Read and mmap the json file */
+  int fd = open(OTA_RESPONSE_FILE_PATH, O_RDONLY);
+  if (fd == -1) {
+    return 1;
+  }
+
+  struct stat sb = {0};
+
+  if (fstat(fd, &sb) == -1) {
+    close(fd);
+    return 1;
+  }
+
+  void *fmap = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+  close(fd);
+  if (fmap == MAP_FAILED) {
+    return 1;
+  }
+
+  /* Parse json into a json_object struct */
+  struct json_object *fjson = json_tokener_parse(fmap);
+
+  struct json_object *json_url = json_object_object_get(fjson, "url");
+  strncpy(parsed_resp->url,
+          json_object_get_string(json_url),
+          json_object_get_string_len(json_url));
+
+  struct json_object *json_version = json_object_object_get(fjson, "version");
+  strncpy(parsed_resp->version,
+          json_object_get_string(json_version),
+          json_object_get_string_len(json_version));
+
+  struct json_object *json_sha256 = json_object_object_get(fjson, "sha256");
+  strncpy(parsed_resp->sha256,
+          json_object_get_string(json_sha256),
+          json_object_get_string_len(json_sha256));
+
+  /* Cleanup objects when done */
+  json_object_object_del(fjson, "");
+
+  return 0;
+}
+
+static int ota_create_files(int *fd_resp, int *fd_img)
+{
+  /* Create empty files */
+  *fd_resp = open(OTA_RESPONSE_FILE_PATH, O_CREAT | O_WRONLY | O_TRUNC, 0644);
+  if (*fd_resp < 0) {
+    piksi_log(LOG_ERR|LOG_SBP, "fd error (%d) \"%s\"", errno, strerror(errno));
+    return 1;
+  }
+
+  *fd_img = open(OTA_IMAGE_FILE_PATH, O_CREAT | O_WRONLY | O_TRUNC, 0644);
+  if (*fd_img < 0) {
+    piksi_log(LOG_ERR|LOG_SBP, "fd error (%d) \"%s\"", errno, strerror(errno));
+    return 1;
+  }
+
+  return 0;
+}
+
+static void ota_close_files(int *fd_resp, int *fd_img)
+{
+  if (*fd_resp >= 0) {
+    close(*fd_resp);
+    *fd_resp = -1;
+  }
+  
+  if (*fd_img >= 0) {
+    close(*fd_img);
+    *fd_img = -1;
+  }
+}
+
+static int ota_client_loop(void)
+{
+  int ret = 0;
+
+  setup_sigterm_handler(ota_sig_handler);
+  setup_sigint_handler(ota_sig_handler);
+
+  /* For OTA_TIMEOUT_S jitter */
+  srand(time(NULL));
+
+  sigwait_params_t params = {0};
+  setup_sigtimedwait(&params, SIGUSR1, OTA_INITIAL_TIMEOUT_S);
+
+  network_context_t *nw_ctx = libnetwork_create(NETWORK_TYPE_OTA);
+
+  int fd_resp = -1;
+  int fd_img = -1;
+
+  while (!libnetwork_shutdown_signaled(nw_ctx) && do_sigtimedwait(&params)) {
```

Comment:
Will `do_sigtimedwait` return when SIGTERM/SIGINT are received?  Otherwise it seems like we won't terminate until `do_sigtimedwait` fires and we have a chance to check `libnetwork_shutdown_signaled`.

Reply:
From https://linux.die.net/man/2/sigtimedwait
```
EINTR

The wait was interrupted by a signal handler; see signal(7). (This handler was for a signal other than one of those in set.)
```

In order for the loop to continue `sigtimedwait` needs to exit with `EAGAIN`.

Also bench testing indicates that behavior is as we expect or want it to be.

Reply:
👍 thanks